src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double MULTIPLIER = 1.5;
+    private static final String MESSAGE_OUT_OF_BOUNDS = "The index passed to any of "
+            + "the methods is invalid";
+    private static final String MESSAGE_NO_ELEMENT = "There is no such element";
+
+    private T[] elements;
+    private int size;
 
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
     }
 
     @Override
-    public void add(T value, int index) {
+    public boolean add(T value) {
+        increaseCapacityIfFull();
+        elements[size++] = (T) value;
+        return true;
+    }
 
+    @Override
+    public void add(T value, int index) {
+        validateIndexForAdd(index);
+        increaseCapacityIfFull();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > elements.length) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T oldValue = (T) elements[index];
+        System.arraycopy(elements,index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int indexOfElement = indexOf(element);
+        if (indexOfElement == -1) {
+            throw new NoSuchElementException(MESSAGE_NO_ELEMENT);
+        }
```

Comment:
1) I think you don't need to create a separate `indexOf()` method because the logic for removing an element isn't big and complicated, so we can iterate over the array here. It's **easier to read/debug** because you don't have to jump to another method just to look and understand how it works
2) Try to create separate methods only if they involve a large chunk of logic or if you use them **at least** 2 times
```suggestion
        for (int i = 0; i < size; i++) {
            if ((element == null && elements[i] == element)
                    || (elements[i] != null && elements[i].equals(element))) {
                return remove(i);
            }
        }
        throw new NoSuchElementException(MESSAGE_NO_ELEMENT);
```